src/create-template-orders.js

```diff
@@ -20,20 +33,263 @@ async function _fetchLastStartTimestamp() {
+        (paymentCreateResponse) => ({
+          action: 'addPayment',
+          payment: {
+            type: 'payment',
+            id: paymentCreateResponse.body.id,
+          },
+        })
+      )
+    }
+    const { body: templateOrder } = await apiRoot
+      .orders()
+      .importOrder()
+      .post({ body: orderDraft })
+      .execute()
+
+    if (addPaymentActions)
+      await _addPaymentWithRetry(templateOrder, addPaymentActions)
+  } catch (err) {
+    if (!isDuplicateOrderError(err)) {
+      const errMsg =
+        `Unexpected error on creating template order from checkout order with number: ${checkoutOrder.orderNumber}.` +
+        ` Line item: ${JSON.stringify(orderDraft.lineItems)}`
+      throw new VError(err, errMsg)
+    }
+  }
+}
+
+async function _setCheckoutOrderProcessed(checkoutOrder) {
+  const updateActions = [
+    {
+      action: 'setCustomField',
+      name: 'isSubscriptionProcessed',
+      value: true,
+    },
+  ]
+  await apiRoot
+    .orders()
+    .withId({ ID: checkoutOrder.id })
+    .post({
+      body: {
+        actions: updateActions,
+        version: checkoutOrder.version,
+      },
+    })
+    .execute()
+}
+
+async function _processCheckoutOrder(checkoutOrder) {
+  try {
+    const templateOrderDrafts = _generateTemplateOrderDrafts(checkoutOrder)
+
+    await pMap(
+      templateOrderDrafts,
+      async (templateOrderDraft) => {
+        await _createTemplateOrderAndPayments(checkoutOrder, templateOrderDraft)
+      },
+      { concurrency: 3 }
+    )
+
+    await _setCheckoutOrderProcessed(checkoutOrder)
+  } catch (err) {
+    logger.error(
+      `Failed to create template order from the checkout order with number ${checkoutOrder.orderNumber}. ` +
+        'Skipping this checkout order' +
+        ` Error: ${JSON.stringify(serializeError(err))}`
+    )
+  }
+}
+
+function isDuplicateOrderError(e) {
+  return (
+    e.code === 400 &&
+    e.body?.errors?.length === 1 &&
+    e.body?.errors?.some(
+      (err) => err.code === 'DuplicateField' && err.field === 'orderNumber'
+    )
+  )
+}
+
+function _generateTemplateOrderDrafts(checkoutOrder) {
+  const orderDrafts = []
+  const subscriptionLineItems = checkoutOrder.lineItems.filter(
+    (lineItem) => lineItem.custom?.fields?.isSubscription
+  )
+  for (const lineItem of subscriptionLineItems)
+    if (lineItem.quantity > 1)
+      for (let i = 0; i < lineItem.quantity; i++) {
+        const templateOrderDraft = _generateTemplateOrderImportDraft(
+          checkoutOrder,
+          lineItem,
+          i + 1
+        )
+        orderDrafts.push(templateOrderDraft)
+      }
+    else {
+      const templateOrderDraft = _generateTemplateOrderImportDraft(
+        checkoutOrder,
+        lineItem
+      )
+      orderDrafts.push(templateOrderDraft)
+    }
+
+  return orderDrafts
+}
+
+function _generateTemplateOrderImportDraft(
+  checkoutOrder,
+  lineItem,
+  quantityIncrement
+) {
+  const schedule = lineItem.custom.fields.schedule
```

Comment:
I think date calculations should be unit tested, so are you thinking to move those into a utility and handle different scenarios with unit tests ?

Reply:
Yes, there is a task for this in the current sprint.